examples/firebase-auth-firestore/app/server/auth.server.ts

```diff
@@ -0,0 +1,48 @@
+import { UserRecord } from "firebase-admin/auth";
+import { signInWithEmailAndPassword } from "firebase/auth";
+import { destroySession, getSession } from "~/sessions";
+import { redirect, Session } from "remix";
+import { auth } from "./firebase.server";
+
+export const checkSessionCookie = async (session: Session) => {
+  try {
+    const decodedIdToken = await auth.server.verifySessionCookie(
+      session.get("session") || ""
+    );
+    return decodedIdToken;
+  } catch {
+    return { uid: undefined };
+  }
+};
+
+export const requireAuth = async (request: Request): Promise<UserRecord> => {
+  const session = await getSession(request.headers.get("cookie"));
+  const { uid } = await checkSessionCookie(session);
+  if (!uid) {
+    throw redirect("/login", {
+      headers: { "Set-Cookie": await destroySession(session) },
+    });
+  }
+  return auth.server.getUser(uid);
+};
+
+export const signIn = async (email: string, password: string) => {
+  const { user } = await signInWithEmailAndPassword(
+    auth.client,
+    email,
+    password
+  );
```

Comment:
I don't think it's safe to use the firebase client sdk on server-side. It is a stateful lib that has the concept of a "current user".

Reply:
From [the Firebase Docs](https://firebase.google.com/docs/web/setup) on the client SDK:

> Follow this guide to use the Firebase JavaScript SDK in your web app or as a client for end-user access, for example, in a Node.js desktop or IoT application.

Of course, you can mess up your server implementation. I don't yet see why receiving the user directly from the sign-in or verify-token calls should be unsafe. See [this comment](https://github.com/firebase/firebase-admin-node/issues/1141#issuecomment-765627139).

Reply:
FWIW I've been dealing with this recently myself. I'm not sure if what I did is correct, but I decided to handle the authentication purely on the client, then I POST the token to the server where it is verified and then set the session cookie. I previously did the auth on the server like this example, but then I could not use the federated logins (like Apple, Facebook, Google, etc). Handling it on the client side was the only way I could get the federated logins to work.

Reply:
It's totally fine to use Firebase on the client and send the JWT through a cookie for the following requests. This example is to show how to use Firebase in your Remix app without using Firebase on the client. From there on you can enhance with client side js

Reply:
> Follow this guide to use the Firebase JavaScript SDK in your web app or as a client for end-user access, for example, in a Node.js desktop or IoT application.


> for end-user access

i.e. not for a server handling requests for multiple users

> in a Node.js desktop

i.e. where there would only be a single user at a time

> IoT application

Similarly, I think they are inferring where you would have a single user at a time.

As @jadlr mentioned, the client SDK is stateful. When you sign in on the server, this user remains signed in for all requests until the next sign in attempt. Sure, it is likely possible to avoid revealing this security hole to end users by limiting access to the API, but I don't think it's a good idea to put this out there as a recommended approach.


> See [this comment](https://github.com/firebase/firebase-admin-node/issues/1141#issuecomment-765627139).

This is a comment saying that they use it in integration tests. I'm not convinced this means it has been built and tested for the purpose of a multi user enviroment.

Reply:
I'm in the process of integrating Firebase auth in to a remix app via the REST API. I'll try report back on how I get on. I have this working via the Firebase emulator at the moment but not in production. I saw [issues with rate limiting](https://twitter.com/1fiws/status/1521426483037839362) being mentioned on Twitter which I think would apply to both the REST and client SDK approaches.

Reply:
@penx thank you for investigating this further. Indeed the firebase sdk's dont seem to be intended to be used this way which may or may not cause risks using this in production. I didn't have the time to investigate the REST APIs. I'm really looking forward to see the results of your investigations!

Reply:
I have auth working with the Firebase REST API now. I plan to work on some Firebase PRs for Remix in the coming weeks, including updating the example to use the REST API.

I think I've also figured a way to get around the rate limiting when used server side so will include that if it works, as far as I'm aware the rate limiting effects both the REST API and the client side SDK when used from the same IP.

My plan is to authenticate on the client via the REST API if the JavaScript has loaded, and fall back to doing it in the action JavaScript has not loaded or failed. Then when the action received a request, it would either:

- contain username/password => fallback to server side login via rest API
- contain response from Firebase after a client side login => server side validate the session before saving in to Remix session (as validation would not be subject to rate limiting)